frontend/src/metabase/admin/permissions/types.ts

```diff
@@ -37,34 +37,83 @@ export type TableEntityId = SchemaEntityId & {
 export type EntityId = DatabaseEntityId &
   Partial<Omit<TableEntityId, "databaseId">>;
 
-export type DataPermission = "data" | "download" | "data-model" | "details";
+export enum DataPermission {
+  VIEW_DATA = "view-data",
+  CREATE_QUERIES = "create-queries",
+  DOWNLOAD = "download",
+  DATA_MODEL = "data-model",
+  DETAILS = "details",
+}
+
+export enum DataPermissionType {
+  ACCESS = "access",
+  NATIVE = "native",
+  DETAILS = "details",
+  DOWNLOAD = "download",
+  DATA_MODEL = "data-model",
+}
+
+export enum DataPermissionValue {
+  BLOCKED = "blocked",
+  CONTROLLED = "controlled",
+  IMPERSONATED = "impersonated",
+  LEGACY_NO_SELF_SERVICE = "legacy-no-self-service",
+  NO = "no",
+  QUERY_BUILDER = "query-builder",
+  QUERY_BUILDER_AND_NATIVE = "query-builder-and-native",
+  SANDBOXED = "sandboxed",
+  UNRESTRICTED = "unrestricted",
+  // download specific values
+  NONE = "none",
+  LIMITED = "limited",
+  FULL = "full",
+  // details specific values
+  YES = "yes",
+  // data model specific values
+  ALL = "all",
+}
 
 export type PermissionSubject = "schemas" | "tables" | "fields";
 
 export type PermissionSectionConfig = {
-  permission: string;
-  type: string;
+  permission: DataPermission;
+  type: DataPermissionType;
+  value: DataPermissionValue;
   isDisabled: boolean;
   disabledTooltip: string | null;
   isHighlighted: boolean;
-  value: string;
-  warning: string | null;
+  warning?: string | null;
   options: {
     label: string;
     value: string;
     icon: string;
     iconColor: string;
   }[];
-  actions: {
-    controlled: {
-      label: string;
-      icon: string;
-      iconColor: string;
-      actionCreator: (...args: unknown[]) => void;
-    }[];
-  };
-  postActions: {
-    controlled: null | ((...args: unknown[]) => void);
-  };
-  confirmations: unknown[];
+  actions?: Partial<
+    Record<
+      DataPermissionValue,
+      | {
+          label: string;
+          icon: string;
+          iconColor: string;
+          actionCreator: (...args: unknown[]) => void;
+        }[]
+      | undefined
```

Comment:
I don't fully understand the context here, but why not use an empty array rather than allow a DataPermissionValue to be mapped to undefined?

Reply:
That's a useful pattern. Given line 100, the values in the actions Record are `Array | undefined` rather than `Record | undefined`. I think `actions.key.x` would raise a similar error even without `| undefined`?